packages/pigeon/lib/swift_generator.dart

```diff
@@ -493,107 +493,111 @@ import FlutterMacOS
                 rootClassNameSet.contains(field.type.baseName)) {
-              toWriteValue = '$fieldName$nullsafe.toMap()';
+              toWriteValue = '$fieldName$nullsafe.toList()';
             } else if (!hostDatatype.isBuiltin &&
                 rootEnumNameSet.contains(field.type.baseName)) {
               toWriteValue = '$fieldName$nullsafe.rawValue';
             } else {
               toWriteValue = field.name;
             }
 
-            final String comma = klass.fields.last == field ? '' : ',';
+            final String comma =
+                getFieldsInSerializationOrder(klass).last == field ? '' : ',';
 
-            indent.writeln('"${field.name}": $toWriteValue$comma');
+            indent.writeln('$toWriteValue$comma');
           }
         });
       });
     }
 
-    void writeFromMap() {
+    void writeFromList() {
       final String className = klass.name;
-      indent
-          .write('static func fromMap(_ map: [String: Any?]) -> $className? ');
+      indent.write('static func fromList(_ list: [Any?]) -> $className? ');
 
       indent.scoped('{', '}', () {
-        for (final NamedType field in klass.fields) {
+        getFieldsInSerializationOrder(klass)
+            .toList()
+            .asMap()
+            .forEach((int index, final NamedType field) {
           final HostDatatype hostDatatype = getHostDatatype(field);
 
-          final String mapValue = 'map["${field.name}"]';
+          final String listValue = 'list[$index]';
           final String fieldType = _swiftTypeForDartType(field.type);
 
           if (field.type.isNullable) {
             if (!hostDatatype.isBuiltin &&
                 rootClassNameSet.contains(field.type.baseName)) {
               indent.writeln('var ${field.name}: $fieldType? = nil');
-              indent.write(
-                  'if let ${field.name}Map = $mapValue as? [String: Any?] ');
+              indent.write('if let ${field.name}List = $listValue as? [Any?] ');
               indent.scoped('{', '}', () {
                 indent.writeln(
-                    '${field.name} = $fieldType.fromMap(${field.name}Map)');
+                    '${field.name} = $fieldType.fromList(${field.name}List)');
               });
             } else if (!hostDatatype.isBuiltin &&
                 rootEnumNameSet.contains(field.type.baseName)) {
               indent.writeln('var ${field.name}: $fieldType? = nil');
-              indent.write('if let ${field.name}RawValue = $mapValue as? Int ');
+              indent
+                  .write('if let ${field.name}RawValue = $listValue as? Int ');
               indent.scoped('{', '}', () {
                 indent.writeln(
                     '${field.name} = $fieldType(rawValue: ${field.name}RawValue)');
               });
             } else {
-              indent.writeln('let ${field.name} = $mapValue as? $fieldType ');
+              indent.writeln('let ${field.name} = $listValue as? $fieldType ');
```

Comment:
We don't need to deal with it now, but looking at the generator output for both Swift and Kotlin I think almost all of the uses of `as?` are not what we want. We should be able to substantially simplify the nullable vs non-nullable handling in these generators by always using force casting (`as!` in Swift, `as` in Kotlin), because just because we *could* assign null to a nullable field if the data on the wire format is garbage doesn't mean we *should*.

I'll file an issue.